rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,326 @@
+  symmetric encryption of the recording data.
+- `Identity` and `Recipient` keys used by `age` to wrap and unwrap file keys.
+  These are an asymmetric `X25519` keypair.
+- HSM/KMS-owned `RSA` keypairs used to wrap the `Identity` key. These allow the
+  HSM or KMS keystore to "own" decryption without requiring proxy or host nodes
+  to make encryption requests to the auth server.
+
+### Key Generation
+
+To simplify integration with different HSM/KMS systems, the keypair used for 
+data encryption through `age` will be a software generated `X25519` keypair.
+For the rest of this section I will refer to the public key as the `Recipient`
+and the private key as the `Identity` in keeping with `age` terminology.
+
+The auth server generating the keys will then use the configured CA keystore to
+generate an `RSA` keypair used to encrypt, or wrap, the `Identity`. Once
+encrypted, the `Recipient`, wrapped `Identity`, and wrapping keypair are added
+as a new entry to the `session_recording_config.status.active_keys` list.
+
+If a new auth server is added to the environment, it will find that there is
+already a configured `X25519` keypair in the session recording config. It will
+check if any active keys are accessible (detailed in the next paragraph) and,
+in the case that there are none, generate a new `RSA` wrapping keypair. The new
+keypair  will be added as an entry to the list of active keys but without a
+wrapped `Identity`. Any other auth server with an active key can inspect the
+new entry, unwrap their own copy of the `Identity`, and wrap it again using
+software encryption and the included public `RSA` key provided by the new auth
+server. The re-wrapped `Identity` is then saved as the wrapped key for the new
+entry and both auth servers will be able to decrypt sessions.
+
+When using KMS keystores, auth servers may share access to the same key. In that
+case, they will also share the same wrapped key which can be identified by the
+key ID. For HSMs integrated over PKCS#11, the auth server's host UUID is
+attached to the key ID and will be used to determine access.
+
+### Decryption and Replay
+
+Because decryption will happen in the auth service before streaming to the
+client, the UX of replaying encrypted recordings is no different from
+unencrypted recordings. The auth server will find and unwrap its active key in
+the `session_recording_config` using either the key's identifier within the KMS
+or the `host_uuid` attached to HSM derived keys. It will use that key to
+decrypt the data on the fly as it's streamed back to the client. This should be
+compatible with all replay clients, including web.
+
+If a recording was encrypted with a rotated key, the auth server will also need
+to search the list of rotated keys to find and unwrap the correct key. Public
+keys are included with their rotated private keys in order to facilitate faster
+searching.
+
+### Key Rotation
```

Comment:
will we ever delete keys from the HSM/KMS? how will we prevent these keys from being deleted by the existing DeleteUnusedKeys function?

Reply:
I think the easiest way would be to fetch the active keys from the `SessionRecordingConfigV2` resource and include them in the list of `allKeysInUse`. That should keep them around and also clean them up automatically when they're rotated out.

Reply:
Good call 🤔 I think we could just use a different label for these keys. That should keep older auth servers from accidentally deleting them and newer auth servers can target keys with both expected labels